platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/ProblemReportingCrossProjectModelAccess.kt

```diff
@@ -242,624 +207,284 @@ class ProblemReportingCrossProjectModelAccess(
             )
         }
 
-        override fun compareTo(other: Project?): Int {
-            return delegate.compareTo(other)
-        }
-
-        override fun getRootDir(): File {
-            return delegate.rootDir
-        }
-
-        @Deprecated("Use layout.buildDirectory instead")
-        override fun getBuildDir(): File {
-            onAccess("buildDir")
-            @Suppress("DEPRECATION")
-            return delegate.buildDir
-        }
-
-        @Deprecated("Use layout.buildDirectory instead")
-        override fun setBuildDir(path: File) {
-            onAccess("buildDir")
-            @Suppress("DEPRECATION")
-            delegate.buildDir = path
-        }
-
-        @Deprecated("Use layout.buildDirectory instead")
-        override fun setBuildDir(path: Any) {
-            onAccess("buildDir")
-            @Suppress("DEPRECATION")
-            delegate.setBuildDir(path)
-        }
-
-        override fun getName(): String {
-            return delegate.name
-        }
-
-        override fun getBuildFile(): File {
-            return delegate.buildFile
-        }
-
-        override fun getDisplayName(): String {
-            return delegate.displayName
-        }
-
-        override fun getDescription(): String? {
-            onAccess("description")
-            return delegate.description
-        }
-
-        override fun setDescription(description: String?) {
-            onAccess("description")
-            delegate.description = description
-        }
-
-        override fun getGroup(): Any {
-            onAccess("group")
-            return delegate.group
-        }
-
-        override fun setGroup(group: Any) {
-            onAccess("group")
-            delegate.group = group
-        }
-
-        override fun getVersion(): Any {
-            onAccess("version")
-            return delegate.version
-        }
-
-        override fun setVersion(version: Any) {
-            onAccess("version")
-            delegate.version = version
-        }
-
-        override fun getStatus(): Any {
-            onAccess("status")
-            return delegate.status
-        }
-
-        override fun getInternalStatus(): Property<Any> {
-            onAccess("internalStatus")
-            return delegate.internalStatus
-        }
-
-        override fun setStatus(status: Any) {
-            onAccess("status")
-            delegate.status = status
-        }
-
-        @AllowUsingApiForExternalUse
-        override fun getChildProjects(): MutableMap<String, Project> {
-            return delegate.childProjects
-        }
+        override fun getParent(): ProjectInternal? =
```

Comment:
🤔  Does it make sense to override these methods, given that we eventually rely on the delegate to use the right `CrossProjectModelAccess`? I'd simply omit these overrides here.

As food for thought, after we merge the `MutableStateAccessAwareProject`, I wonder if it would make sense to remove the "cross project" access methods from the `ProjectInternal`, and instead handle them in the implementations that actually care about cross-project access. Then it would make sense to override these methods and (instead of just delegating them), actually wrap each cross-project access.

Reply:
@alllex It does make sence ot override it, since it's not plain delegation to `super.getProject`, it's `super.getProject(referrer)`

Reply:
True, but so does the implementation of the same methods in the `DefaultProject`.

The important part is that the implementation of the methods that have the extra `referrer` argument also use the cross-project access thingy to wrap the access. And this happens in the `DefaultProject` anyway.

Reply:
> True, but so does the implementation of the same methods in the DefaultProject.

Well, did want to spent much time on it, but the outcome for calling `super.getParent()` and `super.getParent(referrer)` is different.

Consider:
```
// root build.gradle
// assert :a.parent == root
project(':a').parent#isolationViolation
```
In case with overriding `getParent() >> getParent(referrer)` we are coming to `ProjectInternal.wrap` function with `this == referrer` (essentially both of them are `root`), what is leading to returning unwrapped(!) version of `root`. In it's turn, it making call in the example safe - we are accessing in root project mutable state of root project - fair enough, isn't it?

Alternatively, if not override `getParent()` from superclass, than we are coming through `super.getParent(this)`, where `this` is `:a`, what is leading to return wrapped(!) instance after `ProjectInternal.wrap`, which reports a violation thererefore.

[This test](https://github.com/gradle/gradle/blob/master/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy#L847) exercises this behavior and highlighting the issue when missing overriding.

There are a bunch of tests in `IsolatedProjectsAccessFromGroovyDslIntegrationTest` that exercises other overrides behavior as well

Reply:
Thanks for clarifying it. My eyes were looking right through that extra `referrer` argument that adds extra information and is available only in the `ProblemReportingProject`